CRM/MembershipExtras/Job/OfflineAutoRenewal/MultipleInstallmentPlan.php

```diff
@@ -220,22 +213,47 @@ private function updateFieldsLinkingPeriods($currentContributionID, $nextContrib
    * @throws \CiviCRM_API3_Exception
    */
   private function copyRecurringLineItems($currentContribution, $nextContribution) {
-    $recurringLineItems = $this->getRecurringContributionLineItemsToBeRenewed($currentContribution['id']);
-
-    if (count($recurringLineItems)) {
-      foreach ($recurringLineItems as $lineItem) {
-        unset($lineItem['id']);
-        $lineItem['unit_price'] = $this->calculateLineItemUnitPrice($lineItem);
-        $lineItem['line_total'] = MoneyUtilities::roundToCurrencyPrecision($lineItem['unit_price'] * $lineItem['qty']);
-        $lineItem['tax_amount'] = $this->calculateLineItemTaxAmount($lineItem['line_total'], $lineItem['financial_type_id']);
-
-        $newLineItem = civicrm_api3('LineItem', 'create', $lineItem);
-        CRM_MembershipExtras_BAO_ContributionRecurLineItem::create([
-          'contribution_recur_id' => $nextContribution['id'],
-          'line_item_id' => $newLineItem['id'],
-          'start_date' => $this->membershipsStartDate,
-          'auto_renew' => 1,
-        ]);
+    $recurringLineItems = civicrm_api3('ContributionRecurLineItem', 'get', [
+      'sequential' => 1,
+      'contribution_recur_id' => $currentContribution['id'],
+      'auto_renew' => 1,
+      'is_removed' => 0,
+      'api.LineItem.getsingle' => [
+        'id' => '$value.line_item_id',
+        'entity_table' => ['IS NOT NULL' => 1],
+        'entity_id' => ['IS NOT NULL' => 1],
+      ],
+      'options' => ['limit' => 0],
+    ]);
+
+    if (count($recurringLineItems['values'])) {
+      foreach ($recurringLineItems['values'] as $lineItem) {
+
+        $lineItemParams = $lineItem['api.LineItem.getsingle'];
+        $upgradableMembershipTypeId = NULL;
+        if ($this->isMembershipLineItem($lineItemParams)) {
+          $autoUpgradableMembershipChecker = new CRM_MembershipExtras_Service_AutoUpgradableMembershipChecker();
```

Comment:
good point regarding the performance part, but regarding injecting this as a dependency but I think that I disagree,  this class has many other dependencies and it would look awkward and confusing to inject only this one where the rest are used directly. I see that this class and the other two scheduled job classes grew up and each one has many responsibilities and there is an opportunity to separate them into small single-responsible classes, but it is hard to do all of that in one go and that's why the tests I added here will give us more confidence in the future in restructuring and refactoring these parts, but we also need to add more tests too that cover some other important cases too before starting that. 

for now I will try to find a way to address the performance issue part without the need to inject it into the class the constructor .